clarifications in document

kwantam opened the pull request:
This PR fixes and clarifies a few things:

- resolves ambiguity in sign of a sqrt in SvdW high-level procedural description

- remove pairing-friendly draft cite (we can't go to last call with an I-D ref per Nick)

- update variable names in elliptic curve equations (again, this time I think is better) and make corresponding changes in `map_check.sage`

- clean up unused stuff: abs(x) is not used, one of the constants in the optimized sswu impl is not used

- notational consistency throughout

- slightly improved discussion of indifferentiability

- other minor touch-up issues

See next post for discussion of notation

kwantam commented:
I chose variable names for the elliptic curve equations based on existing notation in the document. Here's what I compiled for current notational usage:

| letter | use | letter | use |
|:------:|------|:-------:|------|
| a | Edwards coeff | A | short Weierstrass coeff |
| b | | B | short Weierstrass coeff |
| c | pcode const | C | long Weierstrass coeff (NEW) |
| d | Edwards coeff | D | long Weierstrass coeff (NEW) |
| e | | E | an Elliptic curve |
| f | map fn in suites | F | field |
| g | y^2 = g(x) | G | group |
| h | cofactor | H | hash fn |
| i | pcode counter | I | primitive element in extension field |
| j | pcode counter | J | Montgomery coeff (NEW) |
| k | security parameter | K | Montgomery coeff (NEW) |
| l | | L | length in hash_to_base |
| m | extension degree | M | secondary curve in Ell2Edw |
| n | group order | N | |
| o | | O | |
| p | field characteristic | P | EC point |
| q | field order | Q | EC point |
| r | subgroup order | R | EC point |
| s | | S | |
| t | pcode tmp var | T | |
| u | input to map_to_curve | U | |
| v | Edwards coord | V | Montgomery coord (NEW) |
| w | Edwards coord | W | Montgomery coord (NEW) |
| x | short Weierstrass coord | X | long Weierstrass coord (NEW) |
| y | short Weierstrass coord | Y | long Weierstrass coord (NEW) |
| z | | Z | map_to_curve parameter |

---

Curve equations are now:

- Short Weierstrass: y^2 = x^3 + A * x + B

- Long Weierstrass: Y^2 = X^3 + C * X^2 + D * X

- Montgomery: K * W^2 = V^3 + J * V^2 + V

- twisted Edwards: a * v^2 + w^2 = 1 + d * v^2 * w^2

kwantam commented:
I'd really appreciate a careful look at variable naming. I've done a couple passes and I'll do another one, but more eyes is always better.

armfazh commented:
I'm still reviewing all the changes. one first suggestion will be whether the equation `Long Weierstrass: Y^2 = X^3 + C * X^2 + D * X` can be removed.
I know we added it for making explicit the transition between Montgomery -> W -> twisted Edwards. But, it might be better to provide the direct map, thus we release two variables.

armfazh commented:
although is correct, Montgomery equation looks strange due to the variables used. We are familiar with this document, and somewhat understand thses choices, but I would prefer to use other variables. 
For example, many people use (u,v) -- but u is already taken
(x, y) -- for Weierstrass is ok
(v,w) -- for Edwards looks good

if t is needed, you can use T0,T1,T2.... for temporal variables.

armfazh commented:
Refactoring this document is not a simple task, but we can try to stick more closely to the literature, even the equations be correct.

kwantam commented:
Thanks for the excellent points. I agree, I don't love the weird variable names.

> if t is needed, you can use T0,T1,T2.... for temporal variables.

Or tmp1, tmp2, which is anyway nice and intuitive.

But: does `K * t^2 = s^3 + J * s^2 + s` look better for Montgomery? Maybe... I think we agree that the capital letters are a bit jarring for variables. But then what do we do with long Weierstrass? Still thinking get rid of it? (But I have to admit that X, Y looks kind of OK, even though V, W doesn't.)

Other random thoughts:

- Are you OK with (X, Y) for long Weierstrass, assuming we don't remove it?

- For coefficients, I guess we could do A, B for short Weierstrass, Aw, Bw for long Weierstrass, and Am, Bm for Montgomery. (I'm not proposing Al, Bl because the `l` is easy to confuse with a `1` or an `I`.) Since the document basically doesn't mix short and long Weierstrass at all, this might actually be OK. But on the other hand, I think the potential for confusing (A, B) with (Aw, Bw) is greater than confusing it with (C, D) or (J, K).

- The idea of getting rid of long Weierstrass altogether is really interesting. If I'm understanding your suggestion correctly, this would mean we'd have to modify the Elligator map description. I haven't thought about what that would look like, but probably it would be fine---just change around the constants and add two more multiplications (by 1, in the case of 25519 and 448) to the map.

kwantam commented:
Or maybe just using tmp1, ... and then switching (V, W) -> (s, t) would already be enough of an improvement? Hmm. Not so hard to do...

kwantam commented:
OK, I've tried to resolve the worry about Montgomery coords. It's now back to (s, t), as it is in draft -05.

To make this happen, I renamed temporary variables "tX" to "tvX". I used this rather than "tmpX" because essentially all of our example code is already indented for three-character variable names (e.g., gx1), so this does the least damage. Also, this takes fewer columns, which is nice because we're sometimes squeezed for space.

I also renamed "s" to "z" in the sqrt appendix, and at the same time I made the other changes I suggested in @chris-wood's recent PR. In doing this, I decided that the sample code to compute c1--c5 for Tonelli-Shanks was not useful (too simplistic, adding complexity to the document for no reason; people can look in example code if they need help).

Finally, I added a commit that changes comments from `//` to `#` everywhere. This is because we're already giving essentially Python-ish pseudocode, so we may as well use the corresponding comment char. Also, again, it saves one column, which is helpful in a few spots.

Thoughts on this version? Happy to keep editing if you think there are further clarity issues @chris-wood @armfazh

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -999,15 +1008,17 @@ The following is a brief definition of elliptic curves, with an emphasis on
+Let F be the finite field GF(q) of prime characteristic p > 3.
+(This document does not consider elliptic curves over fields of characteristic 2 or 3.)
+In most cases F is a prime field, so q = p.
+Otherwise, F is a field extension, so q = p^m for an integer m > 1.
```
These break changes done in #193 .

kwantam replied:
Thanks for the heads-up! I made sure to keep #193's changes when rebasing.

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2014,24 +2038,24 @@ Input: u, an element of F.
-18.  e3 = sgn0(u) == sgn0(y)  // Fix sign of y
+18.  e3 = sgn0(u) == sgn0(y)  # Fix sign of y
 19.   y = CMOV(-y, y, e3)
 20. return (x, y)
```
here it should be `(X, Y)` ?

kwantam replied:
:+1: pushed a pass over the document that addresses this.

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2084,31 +2113,42 @@ unambiguously.
 When hashing to a twisted Edwards curve that does not have a standardized
```
> When defining new twisted Edwards curves, a Montgomery equivalent and rational
> map SHOULD be specified, and the sign of the rational map SHOULD be stated
> unambiguously.

According to the last modifications, this paragraph should be updated too with something like this:

> When defining new twisted Edwards curves, a **Weierstrass** equivalent and rational
> map SHOULD be specified, and the sign of the rational map SHOULD be stated
> unambiguously.

kwantam replied:
This is outside the scope of this PR, so I think we should fix it elsewhere---if we do. But:

Note that this section is (and has been since we included it) a discussion of Edwards-to-Montgomery, not Edwards-to-Weierstrass. So what is written here *is* consistent with this section.

On the other hand, you make a good point: maybe this is needlessly indirect! Why not just do Edwards-to-Weierstrass and then use Elligator2 directly? From my perspective, the answer is that most curves will have an Edwards-to-Montgomery map defined, and thus most readers will already be familiar with that map---so talking about it in those terms will be more familiar to the reader.

Also, remember: the reason we do Edwards-to-Montgomery in the first place is so that Edwards and Montgomery hashes are compatible, *and can share an implementation*. So the whole point is, once someone has a hash to a Montgomery curve, they just implement a map to the Edwards and they're done.

Finally, in practice, most Montgomery curves that people use have B = 1, which means that they're already equivalent to the long Weierstrass form---so this distinction is mostly academic. But from my perspective it's cleaner to keep the internal details of the Montgomery map in the Montgomery section, and treat it as a black box here.

If you'd like to keep discussing this, maybe let's move it to its own issue?

armfazh replied:
Move discussion to 195

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2142,10 +2182,10 @@ meeting the requirements in {{rational-map}}.
 
 - map\_to\_curve\_elligator2 is the mapping of {{elligator2}} to the curve M.
-- rational\_map is a function that takes a point (x, y) on M and
+- rational\_map is a function that takes a point (X, Y) on M and
```
Since elligator2 uses a Weierstrass curve defined above, then `M` (which previously referred to a Montgomery curve) got undefined.

kwantam replied:
Yeah, I see what you mean. We could call it E', I suppose.

The fundamental issue is that, as you have pointed out above, the "Montgomery to Edwards" section is inconsistent. Let's fix this after discussing in #195 rather than here, OK?

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -3396,6 +3478,9 @@ This section gives Sage {{SAGE}} scripts used to generate parameters for the map
 ~~~sage
+# Arguments:
+# - F, a field object, e.g., F = GF(2^521 - 1)
+# - A and B, the coefficients of the curve equation y^2 = x^3 + A * x + B
```
@kwantam: Could you please remind me why is required that `g(x)-Z` be irreducible in `find_z_sswu`?

kwantam replied:
Sure thing! It's an attempt to avoid infringing on an existing patent. See [my post](https://mailarchive.ietf.org/arch/msg/cfrg/jV4Wr4fbMKkd4vzsbEhKbous16Y) from the cfrg mailing list.

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2243,24 +2301,24 @@ Operations:
 
 Constants:
-1. c1 = (p + 1) / 4         // Integer arithmetic
+1. c1 = (p + 1) / 4         # Integer arithmetic
```
```suggestion
1. c1 = (q + 1) / 4         # Integer arithmetic
```

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2084,53 +2113,64 @@ unambiguously.
 Implementations MUST detect exceptional cases and return the value
-(v, w) = (0, 1), which is a valid point on all twisted Edwards curves
-given by the equation above.
+(v, w) = (0, 1), which is a valid point (in fact, the identity point {{BBJLP08}})
```
Something shorter
 
```suggestion
(v, w) = (0, 1), which is the identity point 
```

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2084,53 +2113,64 @@ unambiguously.
-(v, w) = (0, 1), which is a valid point on all twisted Edwards curves
-given by the equation above.
+(v, w) = (0, 1), which is a valid point (in fact, the identity point {{BBJLP08}})
+on all twisted Edwards curves given by the equation above.
```
```suggestion
on all twisted Edwards curves.
```

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -3467,84 +3552,80 @@ def find_z_ell2(F):
-3. return s
+1. tv1 = x^c2
+2.   e = (tv1^2) == x
+3.   z = CMOV(tv1 * c1, tv1, e)
```
```suggestion
3. tv2 = tv1 * c1
4.   z = CMOV(tv2, tv1, e)
5. return z
```

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -3467,84 +3552,80 @@ def find_z_ell2(F):
+1. tv1 = x^c2
+2.   e = (tv1^2) == x
+3.   z = CMOV(tv1 * c1, tv1, e)
+3. return z
```
```suggestion
```

armfazh commented on draft-irtf-cfrg-hash-to-curve.md:
```diff
@@ -2204,68 +2249,81 @@ Input: u, an element of F.
 
 Constants:
-1. c1 = (p + 1) / 4         // Integer arithmetic
+1. c1 = (q + 1) / 4         # Integer arithmetic
```
This constant can be omitted since calling to sqrt already includes this constant as q=3 (mod 4).
Replace 
```
6.   Y = gX1^c1
```
by 
```
6.   Y = sqrt(gX1)
```

kwantam replied:
This would be a nice simplification, but I don't think it works in this context.

The reason is, gX1 is not guaranteed to be square, and our sqrt functions are explicitly only defined for square inputs. The point here is that we're *not necessarily* taking the sqrt of gX1---we might instead get back the sqrt of -gX1.

armfazh replied:
oh, I got it.
So using the exponentiation you are preventing (for instance) the following case: someone implemeted sqrt in such a way that always returns error when the sqrt is not found.

kwantam replied:
Yes, exactly right. Nicely put.

kwantam commented:
@chris-wood if you're happy with this I'll merge.

chris-wood commented:
I'm OK with it -- merging away!
